dill/_dill.py

```diff
@@ -876,12 +939,103 @@ def __getattribute__(self, attr):
 
+
+
+# Used to stay compatible with versions of dill whose _create_cell functions
+# do not have a default value.
+# Can be safely replaced removed entirely (replaced by empty tuples for calls to
+# _create_cell) once breaking changes are allowed.
+_CELL_EMPTY = sentinel('_CELL_EMPTY')
+_CELL_EMPTY_shim = BuiltinShim('_CELL_EMPTY', None)
+_CELL_REF_shim = None
+
+
 if PY3:
-    def _create_cell(contents):
-        return (lambda y: contents).__closure__[0]
+    def _create_cell(contents=None):
+        return (lambda: contents).__closure__[0]
+
 else:
-    def _create_cell(contents):
-        return (lambda y: contents).func_closure[0]
+    def _create_cell(contents=None):
+        return (lambda: contents).func_closure[0]
+
+
+if OLD37:
+    if not IS_PYPY and hasattr(ctypes.pythonapi, 'PyCell_Set'):
+        # CPython
+
+        _PyCell_Set = ctypes.pythonapi.PyCell_Set
+
+        def _setattr(object, name, value):
+            if type(object) is CellType and name == 'cell_contents':
+                _PyCell_Set.argtypes = (ctypes.py_object, ctypes.py_object)
+                _PyCell_Set(object, value)
+            else:
+                setattr(object, name, value)
+
+        def _delattr(object, name):
+            if type(object) is CellType and name == 'cell_contents':
+                _PyCell_Set.argtypes = (ctypes.py_object, ctypes.c_void_p)
+                _PyCell_Set(object, None)
+            else:
+                delattr(object, name)
+
+    # General Python (not CPython) up to 3.6 is in a weird case, where it is
+    # possible to pickle recursive cells, but we can't assign directly to the
+    # cell.
+    elif PY3:
+        # Use nonlocal variables to reassign the cell value.
+        # https://stackoverflow.com/a/59276835
+        __nonlocal = ('nonlocal cell',)
+        exec('''def _setattr(cell, name, value):
+            if type(cell) is CellType and name == 'cell_contents':
+                def cell_setter(value):
+                    %s
+                    cell = value # pylint: disable=unused-variable
+                func = FunctionType(cell_setter.__code__, globals(), "", None, (cell,)) # same as cell_setter, but with cell being the cell's contents
+                func(value)
+            else:
+                setattr(cell, name, value)''' % __nonlocal)
+
+        exec('''def _delattr(cell, name):
+            if type(cell) is CellType and name == 'cell_contents':
+                def cell_deleter():
+                    %s
+                    del cell # pylint: disable=unused-variable
+                func = FunctionType(cell_deleter.__code__, globals(), "", None, (cell,)) # same as cell_deleter, but with cell being the cell's contents
+                func()
+            else:
+                delattr(cell, name)''' % __nonlocal)
+
+    else:
+        # Likely PyPy 2.7. Simulate the nonlocal keyword with bytecode
+        # manipulation.
+        def _create_cell(contents=None):
```

Comment:
The logic is a bit odd to define `_create_cell` above, and then to redefine it here.

Reply:
This case will be yanked in the next version when Python 2.7 is dropped anyway, so I just replaced the mainstream one for 2.7 with one that supports empty cell creation. It will be slower for CPython 2.7, but that version should be dead anyway.